File: AdventOfCode_2015/Day05/rust/src/bin/part2.rs
use itertools::Itertools;
use std::fs::File;
use std::io::{self, BufRead, BufReader};

fn contains_non_overlapping_pair(s: &str) -> bool {
    for i in 0..s.len() - 1 {
        for j in i + 2..s.len() - 1 {
            if s[i..i + 2] == s[j..j + 2] {
                return true;
            }
        }
    }
    false
}

fn contains_repeat_with_one_between(s: &str) -> bool {
    s.chars().tuple_windows().any(|(a, _, c)| a == c)
}

fn is_nice(s: &str) -> bool {
    contains_repeat_with_one_between(s) && contains_non_overlapping_pair(s)
}

fn main() -> io::Result<()> {
    let path = "../input.txt";
    let file = File::open(path)?;
    let reader = BufReader::new(file);

    let nice_count = reader
        .lines() // This produces an iterator over Result<String, io::Error>
        .filter_map(|line| match line {
            Ok(l) => Some(l), // If line is Ok, continue
            Err(e) => {
                eprintln!("Error reading line: {}", e);
                None
            }
        })
        .filter(|line| is_nice(line)) // Filter nice strings
        .count(); // Count how many nice strings there are

    println!("Number of nice strings: {}", nice_count);

    Ok(())
}

File: AdventOfCode_2015/Day05/java/Part2.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;


public class Part2 {
    public static void main(String[] args) {
        String filepath = "../input.txt";

        try (BufferedReader reader = new BufferedReader(new FileReader(filepath))) {
            String line;
            int count = 0;
            while ((line = reader.readLine()) != null) {
                if (isNice(line)) {
                    count++;
                }
            }
            System.out.println(count);
        } catch (IOException e) {
            System.err.println("Error reading the input file: " + e.getMessage());
        }
    }

    private static boolean isNice(String line) {
        return hasPairThatAppearsTwice(line) && hasLetterWithOneBetween(line);
    }

    private static boolean hasLetterWithOneBetween(String line) {
        for (int i = 0; i < line.length() - 2; i++) {
            if (line.charAt(i) == line.charAt(i + 2)) {
                return true;
            }
        }
        return false;
    }

    private static boolean hasPairThatAppearsTwice(String line) {
        for (int i = 0; i < line.length() - 1; i++) {
            String pair = line.substring(i, i + 2);

            for (int j = i + 2; j < line.length() - 1; j++) {
                if (line.substring(j, j + 2).equals(pair)) {
                    return true;
                }
            }
        }
        return false;
    }


}
